homeassistant/components/formula_one/sensor.py

```diff
@@ -0,0 +1,419 @@
+
+from .const import DOMAIN, F1_DISCOVERY_NEW, F1_STATE_MULTIPLE
+from .coordinator import F1UpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Configure sensors from a config entry created in the integrations UI."""
+
+    @callback
+    def async_setup_sensors() -> None:
+        _LOGGER.debug("Handling new drivers/constructors/races")
+
+        coordinator: F1UpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+        entities: list[F1Sensor] = []
+
+        for standing in coordinator.constructor_standings:
+            new_cp_sensor = F1ConstructorPositionSensor(coordinator, standing.position)
+            if new_cp_sensor.unique_id not in coordinator.entity_ids:
+                entities.append(new_cp_sensor)
+
+            new_cn_sensor = F1ConstructorNameSensor(
+                coordinator,
+                standing.constructor.constructor_id,
+                standing.constructor.name,
+            )
+            if new_cn_sensor.unique_id not in coordinator.entity_ids:
+                entities.append(new_cn_sensor)
+
+        for standing in coordinator.driver_standings:
+            new_dp_sensor = F1DriverPositionSensor(coordinator, standing.position)
+            if new_dp_sensor.unique_id not in coordinator.entity_ids:
+                entities.append(new_dp_sensor)
+
+            new_dn_sensor = F1DriverNameSensor(
+                coordinator,
+                standing.driver.driver_id,
+            )
+            if new_dn_sensor.unique_id not in coordinator.entity_ids:
+                entities.append(new_dn_sensor)
+
+        for race in coordinator.races:
+            new_race_sensor = F1RaceSensor(coordinator, race.round_no)
+            if new_race_sensor.unique_id not in coordinator.entity_ids:
+                entities.append(new_race_sensor)
+
+        next_race_sensor = F1NextRaceSensor(coordinator)
+        if next_race_sensor.unique_id not in coordinator.entity_ids:
+            entities.append(next_race_sensor)
+
+        async_add_entities(entities)
+
+    async_setup_sensors()
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(hass, F1_DISCOVERY_NEW, async_setup_sensors)
+    )
+
+
+class F1Sensor(CoordinatorEntity):
+    """Base class for a F1 sensor."""
+
+    def _update_value(self) -> None:
+        raise NotImplementedError
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._update_value()
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Handle event when sensor is added."""
+        await super().async_added_to_hass()
+        if self.unique_id is not None:
+            unique_id: str = self.unique_id
+            self.coordinator.entity_ids.append(unique_id)
```

Comment:
We can't do this here, there's a race against `async_setup_sensors`, and the sensor may very well be added twice. Do it in `async_setup_sensors` instead.